subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/OutputUnpacker.java

```diff
@@ -107,6 +109,23 @@ private static void resolveOutputFilePropertySpecs(
         boolean locationOnly,
         Consumer<OutputFilePropertySpec> consumer
     ) {
+        if (filePropertyType == OutputFilePropertyType.DIRECTORIES || filePropertyType == OutputFilePropertyType.FILES) {
+            // Eagerly unpack the value, because we need to create one property per root
+            Object unpackedValue = unpackValue(value, locationOnly);
+            if (unpackedValue == null) {
+                return;
+            }
+            resolveCompositeOutputFilePropertySpecs(ownerDisplayName, propertyName, unpackedValue, filePropertyType.getOutputType(), fileCollectionFactory, consumer);
+        } else {
+            FileCollectionInternal outputFiles = fileCollectionFactory.resolving((Callable<Object>) () -> unpackValue(value, locationOnly));
+            DefaultCacheableOutputFilePropertySpec filePropertySpec = new DefaultCacheableOutputFilePropertySpec(propertyName, null, outputFiles, filePropertyType.getOutputType());
```

Comment:
Is it now somewhere observable that the output files are resolved less often?

Reply:
No, we resolve them the same number of times for now, as evidenced by `TaskParametersIntegrationTest`.

Reply:
Hm.... What does this late resolution mean then? Is there a way to exhibit the behavior in a test?